File: WalletTracker/app/src/main/java/com/chtima/wallettracker/vm/TransactionViewModel.java
package com.chtima.wallettracker.vm;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.LiveDataReactiveStreams;

import com.chtima.wallettracker.dao.repositories.CategoryRepository;
import com.chtima.wallettracker.dao.repositories.TransactionRepository;
import com.chtima.wallettracker.models.Category;
import com.chtima.wallettracker.models.Transaction;

import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class TransactionViewModel extends AndroidViewModel {
    private final TransactionRepository repository;
    private final LiveData<List<Transaction>> liveData;

    public TransactionViewModel(@NonNull Application application) {
        super(application);
        repository = new TransactionRepository(application);
        liveData = LiveDataReactiveStreams.fromPublisher(repository.getAll());
    }

    public LiveData<List<Transaction>> getAll() {
        return liveData;
    }

    public Single<Long> insert(Transaction transaction){
        return repository.insert(transaction)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }
}

File: WalletTracker/app/src/main/java/com/chtima/wallettracker/viewModels/TransactionViewModel.kt
package com.chtima.wallettracker.viewModels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.toLiveData
import com.chtima.wallettracker.db.DatabaseUpdatedEvent
import com.chtima.wallettracker.db.repositories.TransactionRepository
import com.chtima.wallettracker.models.Transaction
import io.reactivex.rxjava3.core.Single
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe

class TransactionViewModel(private var app: Application) : AndroidViewModel(app) {

    private var transactionRepository: TransactionRepository = TransactionRepository(app)

    fun insert(transaction: Transaction) : Single<Long>{
        return transactionRepository.insert(transaction)
    }


}
